### PR TITLE
Add EKS Addon resource

### LINKED ISSUE #863: Implement eks addons
### What resource do you need?
The eks resource in crossplane currently lacks support for [eks-addons](https://docs.aws.amazon.com/eks/latest/userguide/eks-add-ons.html)

### How could Crossplane help solve your problem?
Implement [eks addons](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks#CreateAddonInput) 

### What is your use case?
provision `kube-proxy`, `core-dns` and `vpc-cni` and let AWS manage them.

### Would you be willing to contribute it using [code generator](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md)?
`eks` doesn't use codegen yet.

### PR DESCRIPTION
**Pull request by MisterMX**
### Description of your changes

This PR adds an `Addon` resource for EKS that allows installing EKS addons on a cluster using Crossplane.

Fixes #863 

Referencers are generated with angryjet and mocks using [mockgen](https://github.com/golang/mock).

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Manually and through unit tests

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by MisterMX** [comment]
@goober thanks! 

> However if you for example want to replace the VPC CNI with Cilium you need to manually delete the aws-node, install Cilium and restart the node groups. Is this something that you think should be covered by the schema right away?

Interesting question. I haven't thought about this use case. I guess the best approach would be to create a managed resource after cluster creation and then delete it. I am not sure if the implementation of this PR would leave room for another approach.

**Comment by haarchri** [comment]
@MisterMX can you rebase ? after this i will check the PR thanks

**Comment by MisterMX** [comment]
@haarchri Done. Thanks a lot 👍

File: pkg/clients/eks/fargateprofile.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ import (
 	"github.com/google/go-cmp/cmp/cmpopts"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
-	"github.com/crossplane/provider-aws/apis/eks/v1alpha1"
+	v1alpha1 "github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"
```
Comment (haarchri) [inline pkg/clients/eks/fargateprofile.go:26]: can we use here `manualv1alpha1` without `v1alpha1` and change the code like in identityproviderconfig

`"github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"`

File: pkg/clients/eks/fargateprofile_test.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ import (
 	ekstypes "github.com/aws/aws-sdk-go-v2/service/eks/types"
 	"github.com/google/go-cmp/cmp"
 
-	"github.com/crossplane/provider-aws/apis/eks/v1alpha1"
+	v1alpha1 "github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"
```
Comment (haarchri) [inline pkg/clients/eks/fargateprofile_test.go:29]: can we use here `manualv1alpha1` without `v1alpha1` and change the code like in identityproviderconfig

`"github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"`

File: pkg/clients/eks/nodegroup_test.go [hunk 3]
```diff
@@ -28,7 +28,7 @@ import (
 	"github.com/google/go-cmp/cmp"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
-	"github.com/crossplane/provider-aws/apis/eks/v1alpha1"
+	v1alpha1 "github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"
```
Comment (haarchri) [inline pkg/clients/eks/nodegroup_test.go:31]: can we use here `manualv1alpha1` without `v1alpha1` and change the code like in identityproviderconfig

`"github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"`

File: pkg/controller/eks/addon/setup.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	obj.ClusterName = cr.Spec.ForProvider.ClusterName
+	return nil
+}
+
+func postObserve(_ context.Context, cr *v1alpha1.Addon, _ *awseks.DescribeAddonOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	cr.SetConditions(xpv1.Available())
```
Comment (haarchri) [inline pkg/controller/eks/addon/setup.go:102]: is this the only status we have for AddOns ?

File: pkg/controller/eks/fargateprofile/controller.go [hunk 5]
```diff
@@ -37,7 +37,7 @@ import (
 	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
 	"github.com/crossplane/crossplane-runtime/pkg/resource"
 
-	"github.com/crossplane/provider-aws/apis/eks/v1alpha1"
+	v1alpha1 "github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"
```
Comment (haarchri) [inline pkg/controller/eks/fargateprofile/controller.go:40]: can we use here `manualv1alpha1` without `v1alpha1` and change the code like in identityproviderconfig

`"github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"`

File: pkg/controller/eks/fargateprofile/controller_test.go [hunk 6]
```diff
@@ -32,7 +32,7 @@ import (
 	"github.com/crossplane/crossplane-runtime/pkg/resource"
 	"github.com/crossplane/crossplane-runtime/pkg/test"
 
-	"github.com/crossplane/provider-aws/apis/eks/v1alpha1"
+	v1alpha1 "github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"
```
Comment (haarchri) [inline pkg/controller/eks/fargateprofile/controller_test.go:35]: can we use here `manualv1alpha1` without `v1alpha1` and change the code like in identityproviderconfig

`"github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"`

File: pkg/controller/eks/nodegroup/controller.go [hunk 7]
```diff
@@ -37,7 +37,7 @@ import (
 	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
 	"github.com/crossplane/crossplane-runtime/pkg/resource"
 
-	"github.com/crossplane/provider-aws/apis/eks/v1alpha1"
+	v1alpha1 "github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"
```
Comment (haarchri) [inline pkg/controller/eks/nodegroup/controller.go:40]: can we use here `manualv1alpha1` without `v1alpha1` and change the code like in identityproviderconfig

`"github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"`

File: pkg/controller/eks/nodegroup/controller_test.go [hunk 8]
```diff
@@ -31,7 +31,7 @@ import (
 	"github.com/crossplane/crossplane-runtime/pkg/resource"
 	"github.com/crossplane/crossplane-runtime/pkg/test"
 
-	"github.com/crossplane/provider-aws/apis/eks/v1alpha1"
+	v1alpha1 "github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"
```
Comment (haarchri) [inline pkg/controller/eks/nodegroup/controller_test.go:34]: can we use here `manualv1alpha1` without `v1alpha1` and change the code like in identityproviderconfig

`"github.com/crossplane/provider-aws/apis/eks/manualv1alpha1"`

**Review by haarchri — COMMENTED** [review]
overall Create / Delete for AddOn is working - please check my comments

File: pkg/clients/eks/fargateprofile.go (hunk 1 above)
Comment (MisterMX) [reply pkg/clients/eks/fargateprofile.go]: Fixed

File: pkg/clients/eks/fargateprofile_test.go (hunk 2 above)
Comment (MisterMX) [reply pkg/clients/eks/fargateprofile_test.go]: Fixed

File: pkg/clients/eks/nodegroup_test.go (hunk 3 above)
Comment (MisterMX) [reply pkg/clients/eks/nodegroup_test.go]: Fixed

File: pkg/controller/eks/fargateprofile/controller.go (hunk 5 above)
Comment (MisterMX) [reply pkg/controller/eks/fargateprofile/controller.go]: Fixed

File: pkg/controller/eks/fargateprofile/controller_test.go (hunk 6 above)
Comment (MisterMX) [reply pkg/controller/eks/fargateprofile/controller_test.go]: Fixed

File: pkg/controller/eks/nodegroup/controller.go (hunk 7 above)
Comment (MisterMX) [reply pkg/controller/eks/nodegroup/controller.go]: Fixed

File: pkg/controller/eks/nodegroup/controller_test.go (hunk 8 above)
Comment (MisterMX) [reply pkg/controller/eks/nodegroup/controller_test.go]: Fixed

File: pkg/controller/eks/addon/setup.go (hunk 4 above)
Comment (MisterMX) [reply pkg/controller/eks/addon/setup.go]: Fixed: Added additional conditions for addon status

File: examples/eks/addon.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
+---
+apiVersion: eks.aws.crossplane.io/v1alpha1
+kind: Addon
+metadata:
+  name: sample-addon
+  labels:
+    example: "true"
+spec:
+  forProvider:
+    region: eu-central-1
```
Comment (haarchri) [inline examples/eks/addon.yaml:10]: please change to `us-east-1` that it fits to the clusterRef

File: examples/eks/addon.yaml [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    addon: coredns
+    addonVersion: "1.8.4"
```
Comment (haarchri) [inline examples/eks/addon.yaml:12]: change Version to     `addonVersion: "v1.8.4-eksbuild.1"`

**Review by haarchri — CHANGES_REQUESTED** [review]
please change the example then we are ready to merge - thanks for contribution 

```
Name:         sample-addon
Namespace:    
Labels:       example=true
Annotations:  crossplane.io/external-create-pending: 2021-11-07T10:46:50+01:00
              crossplane.io/external-create-succeeded: 2021-11-07T10:46:52+01:00
              crossplane.io/external-name: arn:aws:eks:us-east-1:255932642927:addon/platform-ref-aws-cluster/coredns/28be7ce5-90c1-0bec-083e-f25b36990e5d
API Version:  eks.aws.crossplane.io/v1alpha1
Kind:         Addon
Metadata:
  Creation Timestamp:  2021-11-07T09:46:24Z
  Finalizers:
    finalizer.managedresource.crossplane.io
  Generation:        3
  Resource Version:  9389000
  Self Link:         /apis/eks.aws.crossplane.io/v1alpha1/addons/sample-addon
  UID:               cbb455e2-1d66-41f7-8d26-b9ed7343bb85
Spec:
  Deletion Policy:  Delete
  For Provider:
    Addon Name:     coredns
    Addon Version:  v1.8.4-eksbuild.1
    Cluster Name:   platform-ref-aws-cluster
    Cluster Name Ref:
      Name:  platform-ref-aws-cluster-g2klb-l7skc
    Region:  us-east-1
    Tags:
      Crossplane - Kind:            addon.eks.aws.crossplane.io
      Crossplane - Name:            sample-addon
      Crossplane - Providerconfig:  default
  Provider Config Ref:
    Name:  default
Status:
  At Provider:
    Addon ARN:     arn:aws:eks:us-east-1:255932642927:addon/platform-ref-aws-cluster/coredns/28be7ce5-90c1-0bec-083e-f25b36990e5d
    Cluster Name:  platform-ref-aws-cluster
    Created At:    2021-11-07T09:46:51Z
    Health:
    Modified At:  2021-11-07T09:47:29Z
    Status:       ACTIVE
  Conditions:
    Last Transition Time:  2021-11-07T09:47:53Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-11-07T09:46:52Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
```

File: examples/eks/addon.yaml (hunk 9 above)
Comment (MisterMX) [reply examples/eks/addon.yaml]: Fixed

File: examples/eks/addon.yaml (hunk 10 above)
Comment (MisterMX) [reply examples/eks/addon.yaml]: Fixed

**Review by haarchri — APPROVED** [review]

**Comment by paul-benetis** [comment]
When I add the `vpc-cni` add-on in the same `Composition` as the cluster, the setup fails.  The add-on installs, but it is not in `Active` state.  I believe it has something to do with the fact that the cluster is still being provisioned when the add-on installation is initialized.  Then, when the add-on install fails as the cluster isn't ready yet, it doesn't heal itself.  I checked the `aws-node` `ServiceAccount` and confirmed that the IAM role ARN was never annotated in the resource.  
This is what the resource looks like in my `Composition`.  Note that `status.cniRoleArn` doesn't get populated until the OIDC provider is created, which depends on the OIDC issuer URL to be available after the cluster is created. 
```
    - name: cni
      base:
        apiVersion: eks.aws.crossplane.io/v1alpha1
        kind: Addon
        spec:
          forProvider:
            addonName: vpc-cni
            clusterNameSelector:
              matchControllerRef: true
      patches:
        - type: PatchSet
          patchSetName: common-parameters
        - type: FromCompositeFieldPath
          fromFieldPath: status.cniRoleArn
          policy:
            fromFieldPath: Required
          toFieldPath: spec.forProvider.serviceAccountRoleARN
        - type: FromCompositeFieldPath
          fromFieldPath: spec.parameters.cniAddonVersion
          toFieldPath: spec.forProvider.addonVersion

**Pull request merged**